Common/OpenCL/Filters/itkGPUInterpolatorCopier.hxx

```diff
@@ -117,29 +118,48 @@ GPUInterpolatorCopier<TTypeList, NDimensions, TInterpolator, TOutputCoordRep>::U
 
     // Try BSpline
     using BSplineInterpolatorType =
-      BSplineInterpolateImageFunction<CPUInputImageType, CPUCoordRepType, CPUCoordRepType>;
+      BSplineInterpolateImageFunction<CPUInputImageType, CPUCoordRepType, double>;
     const typename BSplineInterpolatorType::ConstPointer bspline =
       dynamic_cast<const BSplineInterpolatorType *>(m_InputInterpolator.GetPointer());
 
-    if (bspline)
+    using BSplineInterpolatorFloatType = BSplineInterpolateImageFunction<CPUInputImageType, CPUCoordRepType, float>;
+    const typename BSplineInterpolatorFloatType::ConstPointer bsplineFloat =
+      dynamic_cast<const BSplineInterpolatorFloatType *>(m_InputInterpolator.GetPointer());
+
+    if (bspline || bsplineFloat)
     {
       if (this->m_ExplicitMode)
       {
         // Register image factory because BSplineInterpolateImageFunction
-        // using m_Coefficients as ITK images inside the implementation
+        // using m_Coefficients as ITK images inside the implementation,
+        // and also BSplineDecompositionImageFilter for the calculations
         using GPUImageFactoryType = itk::GPUImageFactory2<TTypeList, NDimensions>;
         using GPUImageFactoryPointer = typename GPUImageFactoryType::Pointer;
         GPUImageFactoryPointer imageFactory = GPUImageFactoryType::New();
         itk::ObjectFactoryBase::RegisterFactory(imageFactory);
 
+        using GPUBSplineDecompositionImageFilterFactoryType = itk::GPUBSplineDecompositionImageFilterFactory2<TTypeList, TTypeList, NDimensions>;
+        using GPUBSplineDecompositionImageFilterFactoryTypePointer = typename GPUBSplineDecompositionImageFilterFactoryType::Pointer;
+        GPUBSplineDecompositionImageFilterFactoryTypePointer decompositionFactory =
+          GPUBSplineDecompositionImageFilterFactoryType::New();
+        itk::ObjectFactoryBase::RegisterFactory(decompositionFactory);
+
         // Create GPU BSpline interpolator in explicit mode
         using GPUBSplineInterpolatorType =
           GPUBSplineInterpolateImageFunction<GPUInputImageType, GPUCoordRepType, GPUCoordRepType>;
         auto bsplineInterpolator = GPUBSplineInterpolatorType::New();
-        bsplineInterpolator->SetSplineOrder(bspline->GetSplineOrder());
-
-        // UnRegister image factory
+        if (bspline)
+        {
+          bsplineInterpolator->SetSplineOrder(bspline->GetSplineOrder());
+        }
+        else
+        {
+          bsplineInterpolator->SetSplineOrder(bsplineFloat->GetSplineOrder());
+        }
```

Comment:
Please consider retrieving the spline order directly after `if (bspline || bsplineFloat)`, for example:

      const auto splineOrder = bspline ? bspline->GetSplineOrder() : bsplineFloat->GetSplineOrder();

So that it can be reused for the non-explicit mode (ExplicitMode == false).